### PR TITLE
Initiate unit tests, less mutability, more keras-like API

### PR DESCRIPTION
**Pull request by maxpumperla**
## What changes were proposed in this pull request?

So, I was *quickly* fixing #29 and one thing led to another... among other things this PR also addresses #9, and does a lot of changes to reshaping, though I'm not sure how much it has to do with #23. More scalaesque, more kerassy.

- [x] Low-level mutable state was replaced by immutable layers
- [x] Companion objects with apply methods for more python-like initialisation
- [x] Made shape inference a little more transparent
- [x] `log` instead of `println`
- [x] Better model definition logging
- [x] First unit test for layers and models
- [x] Eliminated `DenseOutput`, `Dense` has secondary constructor instead.
- [x] Output layers don't take loss functions anymore explicitly in examples, only added in compile step, for more keras resemblance.
- [x] Updated readme to give initial guidance.
- [x] Need more unit tests

## How was this patch tested?

Well, we didn't have unit tests before, at least now they're terrible! Checked for basic shape inference. Have to follow up on this for more coverage.

### CODE REVIEW

**Comment by turambar** [comment]
Excellent word choice: "scalaesque" and "kerassy." Will try to get to this by weekend at latest.

File: src/main/scala/org/deeplearning4j/scalnet/layers/Dense.scala [hunk 1]
```diff
@@ -18,31 +18,56 @@
...
+    val activation: String,
+    val regularizer: WeightRegularizer,
+    val dropOut: Double,
+    override val name: String,
+    lossFunction: Option[LossFunction])
+  extends OutputLayer {
+
+  def this(nOut: Int,
+      nIn: Int = 0,
+      weightInit: WeightInit = WeightInit.XAVIER_UNIFORM,
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/layers/Dense.scala]: Why put the default arguments in this constructor ? Wouldn't they be better placed in the main one, above?

File: src/main/scala/org/deeplearning4j/scalnet/layers/Dense.scala [hunk 2]
```diff
@@ -67,3 +92,18 @@ class Dense(
     }
   }
 }
+
+object Dense {
+
+  def apply(nOut: Int,
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/layers/Dense.scala]: The idiomatic thing would be to have just this one and the main constructor, the Auxiliary constructor you define above seems superfluous.

File: src/main/scala/org/deeplearning4j/scalnet/layers/Node.scala [hunk 3]
```diff
@@ -18,16 +18,27 @@
...
-  def outputShape: List[Int] = _outputShape
+  def name: String
+
+  def inputShape: List[Int]
+
+  def outputShape: List[Int]
+
+  def reshapeInput(nIn: List[Int]): Node = this
+
+  def descibe(): String = "in=" + inputShape + " out=" + outputShape
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/layers/Node.scala]: typo

File: src/main/scala/org/deeplearning4j/scalnet/layers/convolutional/Convolution.scala [hunk 4]
```diff
@@ -31,12 +31,15 @@ abstract class Convolution(
...
-    nChannels: Int = 0,
-    protected val nFilter: Int = 0)
-  extends Node {
-  inputShape = List(nChannels)
-  if (kernelSize.length != stride.length || kernelSize.length != padding.length)
+    protected val nChannels: Int = 0,
+    protected val nIn: Option[List[Int]] = None,
+    protected val nFilter: Int = 0) extends Node {
+
+  override def inputShape: List[Int] = if (nIn.isDefined) nIn.get else List(nChannels)
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/layers/convolutional/Convolution.scala]: a.k.a `nIn.getOrElse(List(nChannels))`

File: src/main/scala/org/deeplearning4j/scalnet/layers/reshaping/Reshape.scala [hunk 5]
```diff
@@ -65,14 +72,22 @@ class Reshape(
   }
 
   private object ReshapePreProcessor{
-    def apply(toShape: Int*): ReshapePreProcessor = new ReshapePreProcessor(null, toShape.toArray, true)
+    def apply(toShape: Int*): ReshapePreProcessor = new ReshapePreProcessor(None, toShape.toArray, true)
   }
 
   override def compile: InputPreProcessor = {
-    if (inputShape.isEmpty || (inputShape.length == 1 && inputShape.head == 0))
+    if (inputShape.isEmpty || (inputShape.length == 1 && inputShape.head == 0)) {
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/layers/reshaping/Reshape.scala]: `PartialFunction.cond(inputShape){ case Nil => True; case 0::Nil => True }`

File: src/main/scala/org/deeplearning4j/scalnet/layers/reshaping/Unflatten3D.scala [hunk 6]
```diff
@@ -34,18 +33,31 @@ import org.deeplearning4j.scalnet.layers.Preprocessor
...
+  override val name = "Unflatten3D"
+
+  override def reshapeInput(newIn: List[Int]): Unflatten3D = {
+    new Unflatten3D(newOutputShape, newIn.head)
+  }
 
-  override def compile: InputPreProcessor = {
-    if (inputShape.isEmpty || (inputShape.length == 1 && inputShape.head == 0))
+    override def compile: InputPreProcessor = {
+    if (inputShape.isEmpty || (inputShape.length == 1 && inputShape.head == 0)) {
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/layers/reshaping/Unflatten3D.scala]: `PartialFunction.cond(inputShape){ case Nil => True; case 0::Nil => True }`

File: src/main/scala/org/deeplearning4j/scalnet/models/Model.scala [hunk 7]
```diff
@@ -18,53 +18,94 @@
...
+  def buildModelConfig(optimizer: Optimizer, seed: Long): NeuralNetConfiguration.Builder = {
+    var builder: NeuralNetConfiguration.Builder = new NeuralNetConfiguration.Builder()
+    if (seed != 0) {
+      builder = builder.seed(seed)
+    }
+    optimizer match {
+      case o: SGD =>
+        builder = builder.optimizationAlgo(o.optimizationAlgorithm)
+          .learningRate(o.lr)
+        if (o.nesterov) {
```
Comment (huitseeker) [inline src/main/scala/org/deeplearning4j/scalnet/models/Model.scala]: Sigh ...

File: src/main/scala/org/deeplearning4j/scalnet/models/Model.scala [hunk 8]
```diff
@@ -18,53 +18,94 @@
...
 
-
 /**
   * Abstract base class for neural net architectures.
   *
   * @author David Kale
   */
 abstract class Model {
+
+  protected var layers: List[Node] = List()
```
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/models/Model.scala]: yep, actually on `Model` level there's much more I want to change. I was mostly focussed on making the lower level stuff immutable first, but thanks for this!

File: src/main/scala/org/deeplearning4j/scalnet/layers/Dense.scala (hunk 2 above)
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/layers/Dense.scala]: no, you're right. I started out with the aux constr. and realised later I need the objects for keras API. Forgot to properly clean up, so to speak.

**Comment by maxpumperla** [comment]
@huitseeker thanks for your review, very interesting comments. As I'm on vacation until 17th, treat this PR however you like. Wait for me, change stuff, merge or delete. :) thanks again

File: src/main/scala/org/deeplearning4j/scalnet/models/Model.scala (hunk 7 above)
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/models/Model.scala]: @huitseeker not sure what precisely you're criticising here, but keep in mind that I didn't even touch this part of the code base. :)  Same with above. I'll try to take care of it regardless.

**Comment by maxpumperla** [comment]
@turambar @huitseeker I'm back in town and ready to go. If you have more comments, please let me know! Thanks

File: src/main/scala/org/deeplearning4j/scalnet/models/Model.scala (hunk 7 above)
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/models/Model.scala]: ok, I did refactor it to this place, but unchanged.

File: src/main/scala/org/deeplearning4j/scalnet/models/Model.scala [hunk 9]
```diff
@@ -18,53 +18,94 @@
...
+  protected var model: MultiLayerNetwork = _
+
+  protected val defaultEpochs = 10
+  protected val defaultOptimizer = SGD(lr = 0.01)
```
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/models/Model.scala]: but `model` is touched in `Sequential` and `NeuralNet`. The other ones I can make a `def`.

Any other suggestion?

File: src/main/scala/org/deeplearning4j/scalnet/models/Model.scala (hunk 8 above)
Comment (maxpumperla) [inline src/main/scala/org/deeplearning4j/scalnet/models/Model.scala]: @huitseeker I don't know how to address this here in a straight-forward manner. If it's OK with you, I'd leave this part as is for now and focus on unit tests instead.

File: README.md [hunk 10]
```diff
@@ -26,3 +26,13 @@ Target for scala-2.10
 ```scala
 $ mvn clean -Pscala-2.10.x
 ```
+
+Having built deeplearning4j and nd4j users can also `sbt` instead of `mvn` to build and test the project, e.g. using
```
Comment (maxpumperla) [inline README.md]: done

**Comment by maxpumperla** [comment]
@agibsonccc @huitseeker @turambar we now have basic unit test coverage for all classes it makes sense to have them. This has already become a huge PR and while I think there's (obviously) still room for improvement, I'd suggest to merge this soon. Feel free to do so!

File: src/test/scala/org/deeplearning4j/scalnet/models/SequentialTest.scala [hunk 11]
```diff
@@ -0,0 +1,81 @@
+/*
```
Comment (maxpumperla) [inline src/test/scala/org/deeplearning4j/scalnet/models/SequentialTest.scala]: done

**Comment by maxpumperla** [comment]
cool, thanks for all the support. Great pleasure to work on this.
